### ISSUE TITLE
Prepare picard-stage for submission

### ISSUE THREAD

**Issue by stuyvenberg** [openspim/micromanager#1]
Here's my to-do list for preparing the DAL. These come from Johannes' comments on each commit to github. This list reflects all noted code as of 9 April (some changes were made before this).

Tests:
- [ ] Thoroughly test all aspects of the device adapter (preferably on the actual system).
- [x] Test DLLMain() removal (other DALs have none -- confirmation?)

Significant/complex:
- [ ] Fix Twister velocity not 'sticking'.
- [x] Get/SetPositionUm aren't guaranteed reversible (rounding effects -- important?)
- [x] Move PiUsb.dll's to PicardStage/ directory
- [x] Fix .vcxproj AdditionalIncludeDirectories

Refactoring:
- [x] Position limits on stage: add properties, constants
- [x] Constants for min/max on XY, step size
- [x] Refactor simple get/set properties to take a member variable reference
- [x] Refactor x/y get/set velocity

Trivial:
- [x] Remove 'built in debug warning'
- [x] Remove parenthesis on ternary operator
- [x] Fix comments on classes
- [x] Ask about copyright
- [x] Use preprocessor TO_STRING
- [x] Remove .vcproj
- [x] Remove pPiDetector initializer

**Comment by dscho** [comment]
> Test DLLMain() removal (other DALs have none -- confirmation?)

``` sh
HOME/micromanager (svn/git-svn) $ git grep DLLMain DeviceAdapters/

```

(Is that answer enough? ;-))

> Fix logging of 'built in debug' warning to use MM::Core (not stdout)

To be precise, we probably want to simply remove that code. It's not useful to the Micro-Manager team (they have a working release process...)

> Move PiUsb.dll's to PicardStage/ directory

Don't worry about that. You'll submit the drivers as .zip file anyway.

> Fix .vcxproj AdditionalIncludeDirectories

I suggested to copy-edit from a working example, e.g. the ASI Stage device adapter.

> Ask about copyright

It should be: Johannes Schindelin, Luke Stuyvenberg Copyright (C) 2011-2014

If you meant about the copyright of the .dll files: that copyright is safely with Picard ;-)

**Comment by stuyvenberg** [comment]
> > Test DLLMain() removal (other DAL's have none -- confirmation?)
> 
> (Is that answer enough? ;-))

Sorry; I was unclear -- I meant, should I take the absence of other instances as confirmation that the routine is unnecessary?

> > Move PiUsb.dll's to PicardStage/ directory
> 
> Don't worry about that. You'll submit the drivers as .zip file anyway.

The other DALs keep their libraries in their DeviceAdapters/ subdirectory... So, comply with the rest of MM, or do as you say?

> > Fix .vcxproj AdditionalIncludeDirectories
> 
> I suggested to copy-edit from a working example, e.g. the ASI Stage device adapter.

The ASI Stage device adapter has no external library dependencies. In particular, unless I separate the PiUsb resources into a header file (moved into DeviceAdapters) and libraries (left in 3rdpartypublic), I can't comply with both this command and the previous.

> > Ask about copyright
> 
> It should be: Johannes Schindelin, Luke Stuyvenberg Copyright (C) 2011-2014

Should it? Are you sure the University of Wisconsin -- Madison, or at least LOCI, doesn't own the copyright to these files? After all, they were for-hire work; looking at all the other DALs seems to suggest a similar policy exists at those institutions...

Additionally, for this file at least, if I _do_ own copyright, I am probably going to forfeit it.

**Comment by dscho** [comment]
> > > Test DLLMain() removal (other DAL's have none -- confirmation?)
> > 
> > (Is that answer enough? ;-))
> 
> Sorry; I was unclear -- I meant, should I take the absence of other instances as confirmation that the routine is unnecessary?

Yes.

;-)

> > > Move PiUsb.dll's to PicardStage/ directory
> > 
> > Don't worry about that. You'll submit the drivers as .zip file anyway.
> 
> The other DALs keep their libraries in their DeviceAdapters/ subdirectory... So, comply with the rest of MM, or do as you say?

We will not be able to provide the .dll within a patch (binary patches are standardized in Git, but we must not assume that upstream uses Git). And as I suspect that the `.vcxproj` file will be re-done anyway, also to allow for the 64-bit drivers, I believe that it does not matter a lot how we link to the `.dll`. What matters more is to get this out of the way (read: to clean it up enough for submission) so that we can continue with the much harder parts of our fork still waiting for us. Every day we spend on the (rather trivial) Picard Stage device adapter patch series is a day we cannot spend on, say, commons-math-3 and more contentious patches in MMCore.

> The ASI Stage device adapter has no external library dependencies.

Of course I meant starting with the `.vcxproj` file of ASI, but then modify it as needed. And since we have an external library dependency, that means we have to add it.

The gist of my suggestion was to try to stay as similar to existing `.vcxproj` files as possible; our `.vcxproj` looks unnecessarily different from ASI Stage's.

> > > Ask about copyright
> > 
> > It should be: Johannes Schindelin, Luke Stuyvenberg Copyright (C) 2011-2014
> 
> Should it?

Yes.

> Are you sure the University of Wisconsin -- Madison, or at least LOCI, doesn't own the copyright to these files? After all, they were for-hire work; looking at all the other DALs seems to suggest a similar policy exists at those institutions...

Sure, it is better to comply with the rest of the source code. When in Rome...

I do care not about the letter of my suggestion. I care about other things, such as being overly precise, to the point of being counter-productive, in the specification of the copyright.  My suggestion was intended to simplify and "undickify" the copyright.

> Additionally, for this file at least, if I do own copyright, I am probably going to forfeit it.

I challenge you to learn more about copyright, and why that statement has no legal standing. But let's do that after submitting the code, and maybe not on LOCI's time.

Speaking of submitting the code: let's really focus on that. I know it was a little too ambitious to hope for half a week of preparation before we can send this upstream, but it is now well over two weeks. I fear we have lost our direction a little bit, and it is time to get back on track.

**Comment by stuyvenberg** [comment]
> And as I suspect that the .vcxproj file will be re-done anyway, also to allow for the 64-bit drivers...

Our project already correctly links the 64-bit drivers; see SHA: 2b941642 (pardon the mass-patch on the .vcproj file; presumably there was a line-ending conflict or somesuch). I will compare the current version to the new download from Picard Industries, and if they differ, I will update our copies and resolve issue 2.

I can't tell for sure that you've answered my question -- 3rdpartypublic, or DeviceAdapters/PicardStage?

> Of course I meant starting with the .vcxproj file of ASI, but then modify it as needed. And since we have an external library dependency, that means we have to add it.

Unfortunately the relevance of this comment was lost with my overhaul of the .vcxproj -- sorry. Does the new version of the file suit your tastes? In particular, to my eye it looks quite similar to that of ASIStage, with some shuffling of element order and a few different properties.

> I challenge you to learn more about copyright, and why that statement has no legal standing.

I did. (Don't worry, I used my own time.) And while I technically can't actually forfeit my copyright (since that would involve giving up my right to revoke transference of rights, if I understood the law correctly), I can accomplish the desired effect by not claiming and never asserting my rights. This was the spirit of my intent.

I appreciate that you intended to simplify the copyright details, but I don't feel comfortable claiming a copyright I may not own. I've e-mailed Kevin (you should have the CC) for his opinion; he will likely know better than either of us.

**Comment by stuyvenberg** [comment]
> Get/SetPositionUm aren't guaranteed reversible (rounding effects -- important?)

We should also decide whether or not this is significant. The intent when I first (and subsequently) edited the `Get/SetPosition` member functions were thus: `Get` should return, as accurately as possible, the present location of the stage, on the same scale and from the same origin as `Set` uses. `Set` should, as accurately as possible, send the stage to the given location.

Currently (and likely always), the position is always positive -- thus, casting a `double` to an `int` performs an effective `floor` operation. So, contingent on correct operation of the stage, `Set` will result in the stage ending at the highest number of steps less than the requested (floating-point) destination.

`Get`, meanwhile, will return the stage's exact step location multiplied by the stored step size property.

I expect, then, that we would see a difference between the intended `Set` and the resultant `Get` -- or equivalently, we would observe `Set(Get())` to cause movement -- only when the machine epsilon of the `Get` causes the output to be at all less than the actual value.

Given the above is true, switching to a proper rounding (i.e. nearest integer instead of `floor`) in `Set` may be desirable.

(Hopefully this is a clear enough explanation -- have I made any mistakes? Do you agree with my conclusion?)

**Comment by dscho** [comment]
>  Get/SetPositionUm aren't guaranteed reversible (rounding effects -- important?)

I will mention that this behavior exists in our code. We should not use that problem as an excuse to delay further, though. (Most likely other device adapters have that problem, too, and the Micro-Manager team probably has a good idea how they want it solved -- or not. We should let them decide.)

**Comment by stuyvenberg** [comment]
> I will mention that this behavior exists in our code.

Alright; I will consider that issue resolved for the moment.

Regarding the location of the third-party (Picard) DLLs, SHA: eb828a009 seems to suggest they should be in the DeviceAdapters/PicardStage directory -- or is that only for non-public DLLs?

(EDIT: I'm beginning to reverse my position on this one -- see, i.e. https://valelab.ucsf.edu/trac/micromanager/browser/DeviceAdapters/HIDManager/HIDManager.vcxproj -- I will, for the moment, use the $(MM_3RDPARTYPUBLIC) variable to link the DLLs.)

Regarding the copyright, I will use a hybrid of the suggestions to date -- our names, followed by the board of regents -- for that line, and our names for the authors line. Note, though, that we may want to indicate MPI's (i.e. PAvel's) role in the development (primarily testing); I will leave the matter of that inclusion up to you.

(EDIT: If you do so wish, Curtis' second and third suggestions provide a meaningful format with which to do so. In the mean time, I see no reason not to use the legally correct information.)

**Comment by stuyvenberg** [comment]
I've pushed several more commits now. They resolve all the checked tasks (to some extent or other).

However, in my initial testing (I have the new stage connected to the dev computer/virtual machine), I noticed a bug in the OnVelocityGeneric refactoring, wherein the twister's velocity won't actually stay set properly. I believe the stage/XY-stage properties hold their value because they are sent to the device and stored in-memory, but the twister only has a 'get' for velocity. I will look into this as time permits.
